pkg/sql/opt/norm/join.go

```diff
@@ -798,3 +798,194 @@ func (c *CustomFuncs) ExtractJoinEquality(
 	// Project away the synthesized columns.
 	return c.f.ConstructProject(join, memo.EmptyProjectionsExpr, leftCols.Union(rightCols))
 }
+func (c *CustomFuncs) CanExtractTupleEqualityLeft(
+	left, right memo.RelExpr,
+	filters memo.FiltersExpr,
+) bool {
+	leftProject, ok := left.(*memo.ProjectExpr)
+	if !ok{
+		return false
+	} 
+	var projectedCols  map[opt.ColumnID]memo.ProjectionsItem
+	var projectedColsSet opt.ColSet
+
+	for _, p := range(leftProject.Projections){
+		if _, isTuple:= p.Element.(*memo.TupleExpr); isTuple {
+			projectedCols[p.Col] = p
+			projectedColsSet.Add(p.Col)
+		}	
+	}
+	hasProjectTupleEquality := false
+	for _, f := range(filters){
+		//we're only interested in filters which reference a projected column
+		if !f.ScalarProps().OuterCols.Intersection(projectedColsSet).Empty() {
+			hasProjectTupleEquality = true
+			if !isValidTupleEqualityExpression(&f, projectedCols){
+				return false
+			}
+		}
+	}
+	return hasProjectTupleEquality
+
+}
+
+func (c *CustomFuncs) CanExtractTupleEqualityRight(
+	left, right memo.RelExpr,
+	filters memo.FiltersExpr,
+) bool {
+	rightProject, ok := right.(*memo.ProjectExpr)
```

Comment:
It is possible that the existing Project, in addition to creating a tuple column, is also projecting away some unused columns from the input. We would still want to keep a Project on the input (with the same passthrough columns) in that case.

Reply:
Ok - does that mean that I should not be trying to hoist the Project above the Join and just focus on splitting out the equalities? That would simplify things, but I'm not sure if the input projections would still prevent conversion to a lookup join.

Reply:
Still not sure what the correct approach should be here?